Adding support for sonata since synapsetool is static

ferdonline opened the pull request:
Building synapsetool statically is used to avoid relying to LD_LIBRARY_PATH (no RPATH in neurodamus)
We then introduce the two synapsetool dependencies directly in neurodamus.

In this patch two aspects were taken into account:
 - Explicitly marking these dependencies `when(~shared)` to prepare for an eventual compat with shared libs in the future
 - We dont rely on synapsetool spec.libs since it was creating problems using external pakcages

ferdonline committed:
Adding support for sonata since synapsetool is static

arsenius7 commented on var/spack/repos/builtin/packages/neurodamus/package.py:
```diff
@@ -87,11 +89,13 @@ def do_stage(self, mirror_only=False):
+            for lib in ['libboost_system-mt', 'libboost_filesystem-mt']:
+                ld_flags += ' ' + find_libraries(lib, spec['boost'].prefix, False, True).ld_flags
+            if spec['synapsetool'].satisfies("+sonata"):
+                ld_flags += ' ' + find_libraries('libsonata', spec['libsonata'].prefix, False, True).ld_flags
```
I tried to do something along these lines, but `ldd`-ing `special` binary exposed that `libsonata` is still dynamically linked (as is Boost, btw).
Is it intended?

ferdonline replied:
The ld_flags end up being generic. e.g.
```
-L/gpfs/bbp.cscs.ch/apps/hpc/spack-deployments/18-06-2018/install/linux-rhel7-x86_64/gcc-6.4.0/boost-1.67.0-6skhmz/lib -lboost_system-mt -L/gpfs/bbp.cscs.ch/apps/hpc/spack-deployments/18-06-2018/install/linux-rhel7-x86_64/gcc-6.4.0/boost-1.67.0-6skhmz/lib -lboost_filesystem-mt -L/gpfs/bbp.cscs.ch/home/leite/test/blueconfigs/INSTALL_HOME/install/linux-rhel7-x86_64/intel-18.0.1/libsonata-develop-fhcp6s/lib -lsonata
```
I wonder if it's a spack bug, of it spack leaves it generic so that the linker could fall back to shared, and for some reason that is happening.

ferdonline replied:
Since we are lookong for static libs anyway what about we use the full path?

pramodk commented:
I will take a look

pramodk commented:
@arsenius7  : do you plan to always install static as well as shared libraries? 

```
kumbhar@bbpv2:/gpfs/bbp.cscs.ch/ssd/tmp_compile$ ll /gpfs/bbp.cscs.ch/ssd/tmp_compile/soft/install/linux-rhel7-x86_64/intel-18.0.1/libsonata-develop-zgztyo/lib/
total 15648
drwxr-xr-x 2 kumbhar bbp     4096 Nov 16 18:48 ./
drwxr-xr-x 5 kumbhar bbp     4096 Nov 16 18:48 ../
-rw-r--r-- 1 kumbhar bbp 11512480 Nov 16 18:48 libsonata.a
-rwxr-xr-x 1 kumbhar bbp  4482896 Nov 16 18:48 libsonata.so*
```

I would suggest to at least add CMake option to not do this i.e. CMake option to disable shared libraries like many other libraries (hdf5, boost etc). We can do alternatives : 
* add `-Wl,-Bstatic` or `-Wl,-Bdynamic` options to prefer specific library
* update recipe to location `libsonata.a` to link

But to integrate well with other packages and "conventions", I prefer to have `shared` variant which allow to build shared/static library with help of adding extra cmake option.

arsenius7 commented:
I don't mind modifying build / install procedure for the sake of consistency, but I'd like to observe that both `libhdf5` and `boost` follow the same practice of placing both static and dynamic versions to `lib` (at least that's what happens for pre-installations in `spack-deployments`).

As a result, as mentioned above, both `libhdf5` and `boost` are also linked dynamically to Neurodamus `special` binary; despite the intent to link `boost` statically in Neurodamus Spack recipe.

ferdonline committed:
Neurodamus link flags to use static libraries

ferdonline commented:
According to docs, it is expected that spack creates generic link flags, (with -L and -l). To ensure we use static libs we can simply provide their full path.

pramodk commented:
> According to docs, it is expected that spack creates generic link flags, (with -L and -l). To ensure we use static libs we can simply provide their full path.

In the end neurodamus doesn't need static if we add `rpath`. Let me fix that today/tomorrow.

pramodk committed:
Add rpath_flags method to LibraryList
    - update synapsetool and neurodamus methods

pramodk commented:
@ferdonline  : could you have a look?

pramodk committed:
Update comment

ferdonline commented on var/spack/repos/builtin/packages/neurodamus/package.py:
```diff
@@ -108,25 +90,26 @@ def build(self, spec, prefix):
+                     spec['reportinglib'].libs.rpath_flags,
                      spec['reportinglib'].libs.ld_flags,
                      spec['hdf5'].prefix.lib,
                      spec['zlib'].prefix.lib)
```
The rpath_flags wasnt to be used with synapsetool/reporting lib only when they satisfy `+shared`?

Btw zlib has been a dep since the beginning, but to be sure, we have mods using zlib directly?

pramodk replied:
>The rpath_flags wasnt to be used with synapsetool/reporting lib only when they satisfy +shared?

yes, its true...but I didn't add it to avoid if condition (adding rpath doesn't hurt anyway)

> Btw zlib has been a dep since the beginning, but to be sure, we have mods using zlib directly?

thats hdf5 dependency...if we link with hdf5, we have to also need to link with zlib and hence that is added as dependency.

ferdonline replied:
I did some tests and shared libs have RPath, so actually we avoid all problems of finding sub-dependencies and things not working when using external packages if we just look for direct dependencies. 
We only need that if we want to support building on static libraries

ferdonline commented on var/spack/repos/builtin/packages/synapsetool/package.py:
```diff
@@ -58,26 +58,41 @@ class Synapsetool(CMakePackage):
+        libraries = find_libraries('libsyn2', root=self.prefix,
+                                  shared=is_shared, recursive=True).libraries
+
+        # boost libraries
```
When syntool is shared shoudnt we avoid reporting dependencies?

pramodk replied:
I don't think syntool static library `contains` dependencies.

ferdonline replied:
The lines spec['boost'] etc really must be conditional, otherwise the top-level package had to find himself all sub-dependencies. As noted above, RPATH is correctly set in shared libs, so we can rely on that to simplify dependency specification to only direct dependencies. I will make some more tests there

ferdonline commented on var/spack/repos/builtin/packages/neurodamus/package.py:
```diff
@@ -108,25 +90,26 @@ def build(self, spec, prefix):
         if '+syntool' in spec:
             include_flag += ' -DENABLE_SYNTOOL -I ' + spec['synapsetool'].prefix.include
-            link_flag += self.syntool_dep_libs()
+            link_flag += spec['synapsetool'].libs.rpath_flags
```
Let me test if this is not a problem if synapsetool is an external package. Btw, any chance we use Spack chain in the near future?

pramodk replied:
> Let me test if this is not a problem if synapsetool is an external package. Btw, any chance we use Spack chain in the near future?

yes! i now remember your comment! if we specify synapse-tool as external package then we have to go with adding all dependencies in neurodamus (which is `not that bad`).

ferdonline committed:
Adding more variants to manually control dependend

...while spack doesnt allow optional dependencies upon other deps
Tested with/without external package

ferdonline committed:
Fixed and tested. Working with sonata, shared/static, external syntool

ferdonline commented:
@pramodk I think I tested all combinations (static/shared, built/external, syn2+sonata/once without), against syntool with sonata change (https://bbpcode.epfl.ch/code/#/c/42930/)
Unless we want to launch a jobs with all possible combinations, I'd say we are good for the moment. But please let me know if you see any issue.

pramodk commented:
I am bit worried about `-Wl,--as-needed` but lets see if/when that causes problem.
